### PR TITLE
[WIP] Cleanup build script

### PR DESCRIPTION
**Pull request by florian-beetz**
This prepares for splitting JabRef into several modules.

The dependencies were extracted to a separate file. This makes it much easier in the future to define shared dependencies across multiple modules, without duplicating them in the build scripts and especially keeping them all updated to the same version.

According to Gradle's guide on [writing maintainable build scripts](https://docs.gradle.org/current/userguide/authoring_maintainable_build_scripts.html), build scripts should not contain any imperative logic but just declarative language elements. Imperative logic should live as plugins in the `buildSrc` directory. So I extracted the ANTLR tasks to this directory. This also makes it much easier to move the tasks to another module in the future.

What do you think? I would like to extract the XJC-, release-, and localization-tasks to separate plugins if you approve these changes.

----

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [x] Manually tested changed features in running JabRef
- [ ] Screenshots added in PR description (for bigger UI changes)
- [ ] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

### CODE REVIEW

File: buildSrc/build.gradle [hunk 1]
```diff
@@ -0,0 +1,13 @@
+plugins {
+    id 'java'
+    id 'groovy'
+}
+
+sourceCompatibility = 1.8
+
+repositories {
+    mavenCentral()
+}
```
Comment (Siedlerchr) [inline buildSrc/build.gradle]: Isn't this now duplicated?

Comment (florian-beetz) [reply buildSrc/build.gradle]: This specifies the repositories used for resolving dependencies for the buildSrc project.
As far as I know, this does not overwrite or duplicate the `buildscript.repositories` property in the `build.gradle`.

**Comment by florian-beetz** [comment]
@Siedlerchr The `buildSrc` code is put on the classpath of the build scripts by gradle, including sub-modules. So in those sub-modules, you can just `import static Dependencies.libraries` and use them as in the main script. Same goes for the tasks I defined.

The main advantage of this is that the versions of dependencies don't need to be managed in every module, but you can just change the version in one central file and all modules will use that version.

File: build.gradle [hunk 2]
```diff
@@ -75,102 +69,81 @@ repositories {
...
-    compile 'javax.inject:javax.inject:1'
+    compile libraries.fontawesome
+    compile libraries.mvvmfx
+    compile libraries.easybind
+    compile libraries.flowless
+    compile libraries.richtextfx
+    compile libraries.dndtabpane
+    compile libraries.javaInject
 
     // Cannot be updated to 9.*.* until Jabref works with Java 9
```
Comment (stefan-kolb) [inline build.gradle]: Move comment to library file?

File: build.gradle [hunk 3]
```diff
@@ -75,102 +69,81 @@ repositories {
...
     // >1.8.0-beta is required for java 9 compatibility
-    compile 'org.slf4j:slf4j-api:1.8.0-beta2'
-    compile 'org.apache.logging.log4j:log4j-slf4j-impl:2.11.0'
-    compile 'org.apache.logging.log4j:log4j-jcl:2.11.0'
-    compile 'org.apache.logging.log4j:log4j-api:2.11.0'
-    compile 'org.apache.logging.log4j:log4j-core:2.11.0'
+    compile libraries.loggingApi
+    compile libraries.logging
 
     // need to use snapshots as the stable version is from 2013 and doesn't support v1.0.1 CitationStyles
```
Comment (stefan-kolb) [inline build.gradle]: Comment to library file

File: buildSrc/build.gradle [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+dependencies {
+}
```
Comment (stefan-kolb) [inline buildSrc/build.gradle]: Do we need this if it's empty?

**Comment by florian-beetz** [comment]
@stefan-kolb I agree that the dependency extraction is not optimal. I think something like mavens `dependencyManagement` would be much better, but afaik Gradle has no such thing.

As far as I can tell, shared dependencies are mainly testing, logging, guava, and antlr.
I guess when JDK 11 is released without JavaFX, it will become a shared dependency as well because the collections are used basically everywhere.

**Comment by stefan-kolb** [comment]
The question is, if we should just include the plugin idea for now and duplicate the dependencies.
After the module split, we could evaluate if it is reasonable to extract the dependencies (of course we should keep the code you wrote anyway).
I'm not sure which way to go.

**Comment by florian-beetz** [comment]
Sure, we can do that.

Just for completeness sake: I've read about two other possibilities to manage dependency versions in multi-module gradle builds. 
One managed the versions in a [`gradle.properties`](https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_configuration_properties) like `versionGuava=25.1-jre`, and kept the dependencies in the build scripts but making the versions variable like `compile "com.google.guava:guava:${versionGuava}"`.
The other is using [this gradle plugin](https://spring.io/blog/2015/02/23/better-dependency-management-for-gradle) together with a Maven BOM, but IMO using the Maven XML with Gradle is pretty ugly.

**Comment by stefan-kolb** [comment]
Using a properties file looks promising to me.
This also would separate code and version declarations pretty well.

**Comment by tobiasdiez** [comment]
The extraction of plugins is definitely a good idea.
In principle, I also like the refactored dependencies except that it is now relatively complex to add a new dependency: you have to add an entry under version, under libraries and in the main gradle file. What do you think about combining the version and libraries lists, i.e. define the libraries in the usual full gradle format like `org.apache.logging.log4j:log4j-core:2.11.0`. Then adding a new dependency is still straightforward but you also have the advantage to reuse it in other modules.

**Comment by tobiasdiez** [comment]
We discussed this in the dev call, and the decision was to include the changes of the plugins (done in https://github.com/JabRef/jabref/pull/4934) and to wait with the extraction of the dependencies until we have a bit more experience with the modular build.

Thanks for your PR.

**Pull request closed without merging**